Draft: Feature/crkrenn/frequency distribution

crkrenn opened the pull request:
@MainRo (cc: @carsonfarmer),

This pull request combines several related changes: 

1. (the inspiration for the pull request): an implementation of a `frequency_density_distribution` function.
2. implementation of a consistent output format for `distogram.histogram`, `distogram.frequency_density_distribution`, and `numpy.histogram`
3. a modification of `_linspace` to ensure that `h.max` is a constant and is not affected by rounding errors
4. a modification of `count_at` to return appropriate values at `h.min` and `h.max`
5. the addition of two tests
6. an update to `distogram.ipynb` that demonstrates the performance and utility of the  `frequency_density_distribution` function.

Please let me know if you have any questions. 

Very gratefully and sincerely, 

-Chris

crkrenn committed:
Update README.rst

Commit:
added

'

Commit:
updated histogram format; updated tests

Commit:
reverted README

Commit:
simplified _linspace

crkrenn commented:
Thanks Carson!

MainRo commented:
Thanks @crkrenn, that will be a neat addition!

I started to review the changes, but not completely yet. 

The multiple output formats add complexity that I would avoid. I understand that your aim was to unify it with the numpy *histogram* output format. On my side, the *distogram* format choice was driven for easy conversions to pandas dataframes. We definitely need to use a single output format. I will think more about this, but introducing a breaking change to align with numpy looks like a possible option. We could add a helper API to ease conversions to pandas dataframes.

In the notebook I think there is a typo in the "Frequency density results retain more detail when there are extreme outliers" section: The "compare with real histogram of the distribution" part uses distogram instead of numpy to compute the histogram.

Also on the notebook, I would remove the last part with the screenshot: The notebook is to play with the API, so this part is not necessary (no need to justify when each API should be used).

crkrenn commented:
@MainRo, Your comments are all good. I knew the example in the notebook wasn't really necessary, but I wanted to give you some context. I'll make your suggested changes soon.

Commit:
updated text in notebook to be consistent between three histograms; removed example'

crkrenn commented:
@MainRo, 

I think that I've responded to your two suggestions about the notebook. 

Previously, I also updated your histogram implementation to remove an "off-by-one" bug/feature. You can verify the update with very small distributions.

Commit:
fixed axes labels

MainRo commented on distogram/__init__.py:
```diff
@@ -326,31 +334,81 @@ def stddev(h: Distogram) -> float:
+def histogram(
+    h: Distogram, 
+    bin_count: int = 100, 
+    data_type: str = "distogram") -> Union[List[Tuple[float, float]],
```
Let's change the result to match the numpy format, so please remove this parameter, and always return numpy format.

crkrenn replied:
I've made the changes to the numpy histogram format. I still want to convince myself that there is not an off by one error in the binning.

MainRo commented on distogram/__init__.py:
```diff
@@ -326,31 +334,81 @@ def stddev(h: Distogram) -> float:
+        raise ValueError(
+            f"data_type {data_type} is not a member of {allowed_data_types}")
+
+    bin_bounds = _linspace(h.min, h.max, num=bin_count + 2)
```
The +2 should be removed to output bin_count+1 bound values. moving to NumPy's format allows removing this inconsistency with the current output.

MainRo reviewed (changes requested):
That's all good, thank you very much!

Let's keep only the numpy's format for consistency, even if this breaks compatibility.

MainRo commented:
If you are short on free time, I can merge the PR as-is. I can then do the remaining changes.

Commit:
converted to numpy format for histogram

Commit:
renamed bins -> binned_data and bin_count -> bins

Commit:
finished converting to numpy histogram format

MainRo commented:
> Can I bump the version to 2.1.0 to indicate the functional change, or does it need to be 3.0.0 because it is not backwards compatible?

Do not change the versions, this will be done when the next release will be published, in another dedicated commit.

MainRo commented:
> PS. I'm also thinking about the pros and cons of changing bin_count to bins in the histogram method to match the numpy arguments.

For consistency within distogram, I prefer to keep bin_count here: This naming is used as a class field, and "bins" is used to store the actual bins of the dynamic histogram. So it may bring some confusion.

Commit:
fixed some off-by-one bugs

Commit:
fixed error in histogram test

crkrenn commented:
I'm sorry I missed your comments above while I worked. 

I fixed some off by one bugs so that the integral of distogram bins or density distributions are the same as the expected (and numpy) result. 

I still think that bumping the version is important, particularly if you plan to push to pypi. If people are using the old version of distogram, they should have the option of still using it without revising their code. According to the [semantic versioning standards](https://py-pkgs.org/07-releasing-versioning.html), making a change that is not backwards compatible should result in a major version increments (e.g. for distogram from 2.X.X to 3.X.X.)

I also think that switching from `bin_count` to `bins` is the right thing to do. I agree that `bin_count` is more clear, but many more people have learned the numpy conventions than have used your package. In the future, someone might also want to support unevenly spaced bins with `distogram.histogram`. 

I can definitely revert the changes, but I did want to share my reasoning. 

Very respectfully and gratefully, 

-Chris

MainRo commented:
> I still think that bumping the version is important

Yes, I will bump the major, but in another commit. This PR is an improvement commit, not a release commit. I never bump a version in development commits. Before a release is done, there will be a dedicated commit to only bump the version accordingly (I do this in a semi-automated way this bumpversion). See 64a237b5d554d73a3d0db50209d3dc557c6e6a8a for example in the last release.
So you do not have to change versions here.

> I also think that switching from `bin_count` to `bins` is the right thing to do. I agree that `bin_count` is more clear, but many more people have learned the numpy conventions than have used your package. 

Sure, but the important point to me is that they (numpy/distogram bin parameters) do not have the same meaning: In distogram it can only be the number of bins, on numpy it can be 3 different things.

>In the future, someone might also want to support unevenly spaced bins with `distogram.histogram`.

If other kinds of binning are supported in the future I think I will prefer adding dedicated parameters than having a single argument with many possible types and meanings.

So I understand your point of being consistent with numpy, but for this parameter, I still would like to keep the current naming :)

crkrenn commented:
All good. Will tweak when I can (with less delay this time).

Commit:
reverted version and bin_count

crkrenn commented:
I've made the changes you requested. 

I also added a text only version of `distogram.ipynb`. To avoid conflict with `import distogram`, I also renamed the notebook to `distogram_notebook.ipynb`. 

[version-control-with-jupyter-notebook](https://towardsdatascience.com/version-control-with-jupyter-notebook-b9630bc5996e)

If you prefer, I can split the notebook change into a separate pull request.

MainRo commented:
Thanks a lot for the new feature and your patience!
The text format conversion of a notebook is great, I did not know that tool.
